tests/suites/test_suite_pk.function

```diff
@@ -17,9 +17,62 @@
  * unconditionally (https://github.com/Mbed-TLS/mbedtls/issues/2023). */
 #include "psa/crypto.h"
 
+/* Used for properly sizing the key buffer in pk_genkey_ec() */
+#include "mbedtls/psa_util.h"
+
 #define RSA_KEY_SIZE 512
 #define RSA_KEY_LEN   64
 
+#if defined(MBEDTLS_USE_PSA_CRYPTO) && defined(MBEDTLS_ECP_C)
+static int pk_genkey_ec(mbedtls_ecp_group *grp,
+                        mbedtls_mpi *d, mbedtls_ecp_point *Q)
+{
+    psa_status_t status;
+    psa_key_attributes_t key_attr = PSA_KEY_ATTRIBUTES_INIT;
+    mbedtls_svc_key_id_t key_id = MBEDTLS_SVC_KEY_ID_INIT;
+    size_t curve_bits;
+    psa_ecc_family_t curve = mbedtls_ecc_group_to_psa(grp->id,
+                                                      &curve_bits);
+    unsigned char key_buf[MBEDTLS_PSA_MAX_EC_PUBKEY_LENGTH];
+    size_t key_len;
+    int ret;
+
+    psa_set_key_type(&key_attr, PSA_KEY_TYPE_ECC_KEY_PAIR(curve));
+    psa_set_key_bits(&key_attr, curve_bits);
+    psa_set_key_usage_flags(&key_attr, PSA_KEY_USAGE_EXPORT);
+
+    status = psa_generate_key(&key_attr, &key_id);
+    if (status != PSA_SUCCESS) {
+        return MBEDTLS_ERR_PK_FEATURE_UNAVAILABLE;
```

Comment:
This is only test but I think we should also destroy key if key generation fails in the middle.

Reply:
it seems to me that this is the only point in that function for which there is a `return` instead of `goto exit`. However this seems ok to me since if `psa_generate_key` fails then there should be nothing to destroy. Other failures destroy the key. What am I missing?

Reply:
Yes, that is all true. I'm not sure if we shouldn't also destroy key here. But looking at the description of `psa_generate_key`:

>On success, an identifier for the newly created key.

It seems that current version is correct.

Reply:
Hmm, I would expect `psa_generate_key()` to act in an atomic way: either it succeeds, or it leave the key slot unchanged, so that callers don't need to call destroy on failure.  However I checked [the documentation](https://armmbed.github.io/mbed-crypto/html/api/keys/management.html#c.psa_generate_key) and it doesn't say explicitly. I checked other uses in the library, and they seem to only call destroy when generate succeeded but an error occurred later. So I think the code is fine as it is, but out of precaution I'll ask on the PSA Crypto API channel.

Reply:
> However I checked [the documentation](https://armmbed.github.io/mbed-crypto/html/api/keys/management.html#c.psa_generate_key) and it doesn't say explicitly.

Well, I apparently didn't get enough sleep last night, because the documentation does say: "key On success, an identifier for the newly created key. [PSA_KEY_ID_NULL](https://armmbed.github.io/mbed-crypto/html/api/keys/ids.html#c.PSA_KEY_ID_NULL) on failure." So, we don't need to call `destroy()` on failure here. (It would however be legal, as `psa_destroy_key(PSA_KEY_ID_NULL)` is guaranteed to do nothing an return success.)

Reply:
It's a general principle that a PSA function doesn't affect the _system_ state on failure. (It might leave its in-out arguments in a changed state on failure, however: multipart operation functions put the operation object in an error state when they fail.)

`psa_generate_key` and the other key creation functions are guaranteed to set the `key_id` output argument to 0 on failure. So it's guaranteed that calling `psa_destroy_key(key_id)` is always correct, but it's also guaranteed that it's a no-op if the creation function fails.

Reply:
Well, I should refresh the page more often when replying to comments :) I hadn't seen your replies when I wrote mines.